python/dolfinx/fem/function.py

```diff
@@ -104,67 +106,56 @@ def __init__(self,
+        constants = [constant._cpp_object for i, constant in enumerate(ufl_constants)]
 
-        # Getcpp Expression type
-        def expressiontype(dtype):
+        def create_expression(dtype):
             if dtype is np.float64:
-                return _cpp.fem.Expression_float64
+                return _cpp.fem.create_expression_float64
             elif dtype is np.complex128:
-                return _cpp.fem.Expression_complex128
+                return _cpp.fem.create_expression_complex128
             else:
                 raise NotImplementedError(f"Type {dtype} not supported.")
 
-        self._cpp_object = expressiontype(dtype)(coefficients, constants, mesh, x, fn, value_size)
+        self._cpp_object = create_expression(dtype)(ffi.cast("uintptr_t", ffi.addressof(self._ufc_expression)),
+                                                    coeffs, constants, mesh)
 
-    def eval(self, cells: np.ndarray, u: typing.Optional[np.ndarray] = None) -> np.ndarray:
+    def eval(self, cells: np.ndarray, values: np.ndarray = None) -> np.ndarray:
         """Evaluate Expression in cells.
 
         Parameters
         ----------
         cells
             local indices of cells to evaluate expression.
-        u: optional
-            array of shape (num_cells, num_points*value_size) to
-            store result of expression evaluation.
 
         Returns
         -------
 
         u: np.ndarray
             The i-th row of u contains the expression evaluated on cells[i].
 
-        Note
-        ----
-        This function allocates u of the appropriate size if u is not passed.
         """
-        cells = np.asarray(cells, dtype=np.int32)
-        assert cells.ndim == 1
-        num_cells = cells.shape[0]
+        _cells = np.asarray(cells, dtype=np.int32)
 
         # Allocate memory for result if u was not provided
-        if u is None:
-            if np.issubdtype(PETSc.ScalarType, np.complexfloating):
-                u = np.empty((num_cells, self.num_points * self.value_size), dtype=np.complex128)
-            else:
-                u = np.empty((num_cells, self.num_points * self.value_size), dtype=np.float64)
-            self._cpp_object.eval(cells, u)
+        if values is None:
+            num_all_argument_dofs = np.prod(self.num_argument_dofs, dtype=int)
+            values = np.empty((_cells.shape[0],
+                               self.num_points * self.value_size * num_all_argument_dofs), dtype=self.dtype)
+            self._cpp_object.eval(cells, values)
         else:
-            assert u.ndim < 3
-            assert u.size == num_cells * self.num_points * self.value_size
-            assert u.shape[0] == num_cells
-            assert u.shape[1] == self.num_points * self.value_size
-            self._cpp_object.eval(cells, u)
+            if values.ndim >= 3 or values.shape[0] != _cells.shape[0] or values.shape[1] != _cells.shape[1]:
+                raise TypeError("Provided vaules array has incorrect shape.")
```

Comment:
What would be the most convenient shape for the caller across the Python interface?